1. Reading from the original dataset

In [ ]:
import pandas as pd
import numpy as np
df= pd.read_excel('data.xlsx')
df.head(2)

2. Filtering out the non US startups

In [ ]:
df_final=df[(df.iloc[:]['headquarters'].str.contains("United States"))]

In [ ]:
print(df_final.head())
df_final.reset_index(inplace= True)
df_final.drop(labels='index', axis=1,inplace= True)
df_final.head(30), len(df_final)

             company_name  ...         filter
3   1 Accord Technologies  ...  Software-2018
5           1005 Ventures  ...  Software-2018
7     1533 Systems, Inc.   ...  Software-2018
9            17 TeraWatts  ...  Software-2018
17            360 Privacy  ...  Software-2018

[5 rows x 11 columns]


/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


(             company_name  ...         filter
 0   1 Accord Technologies  ...  Software-2018
 1           1005 Ventures  ...  Software-2018
 2     1533 Systems, Inc.   ...  Software-2018
 3            17 TeraWatts  ...  Software-2018
 4             360 Privacy  ...  Software-2018
 5                    3DIQ  ...  Software-2018
 6          3P Touch Media  ...  Software-2018
 7                     4Di  ...  Software-2018
 8                    ABYA  ...  Software-2018
 9          Accelitude LLC  ...  Software-2018
 10                Accelor  ...  Software-2018
 11       Accounting Panel  ...  Software-2018
 12      ACE IoT Solutions  ...  Software-2018
 13         Acsibi Studios  ...  Software-2018
 14                  ACTIO  ...  Software-2018
 15          Activeconnect  ...  Software-2018
 16             Acuity PPM  ...  Software-2018
 17          ACW Analytics  ...  Software-2018
 18        Ad Fontes Media  ...  Software-2018
 19              Ad Genius  ...  Software-2018
 20          

3. Dropping those rows which contain null or vague small description 

In [ ]:
drop=[]
for i in range(len(df_final)):
  #print(df_final.iloc[i]['description'])
  if len(df_final.iloc[i]['description'])<5:
    drop.append(i)
    


In [ ]:
df_final.drop(drop, axis=0, inplace= True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
df_final.reset_index(inplace= True)
df_final.drop(labels='index', axis=1,inplace= True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
df_final.head()

company_name  ...         filter
0  1 Accord Technologies  ...  Software-2018
1          1005 Ventures  ...  Software-2018
2    1533 Systems, Inc.   ...  Software-2018
3           17 TeraWatts  ...  Software-2018
4                   3DIQ  ...  Software-2018

[5 rows x 11 columns]

4. Converting excel formatted datetime to number of employees in string format

In [ ]:
import datetime
for i,date in enumerate(df_final['num_employees']):
  if isinstance(date, datetime.datetime):
    #print(df_final.iloc[i]['num_employees'])
    df_final.iloc[i]['num_employees']=date.strftime('%m/%d/%Y')

In [ ]:
#in the num employee column, replace -,/ with ' to ' and remove year

for i,date in enumerate(df_final['num_employees']):
   if '/' in date:# or '-' in date:
     df_final.iloc[i]['num_employees']=date.replace('/', ' to ')
   if '-' in date:
     df_final.iloc[i]['num_employees']=date.replace("-", " to ")

for i,date in enumerate(df_final['num_employees']):
   if "to 2021" in date:
     df_final.iloc[i]['num_employees']=date.replace('to 2021', '')

In [ ]:
df_final.describe()

company_name  ...         filter
count          2996  ...           2996
unique         2787  ...              5
top          Sweave  ...  software_2017
freq              3  ...           1344

[4 rows x 11 columns]

In [ ]:
df_final.head()

company_name  ...         filter
0  1 Accord Technologies  ...  Software-2018
1          1005 Ventures  ...  Software-2018
2    1533 Systems, Inc.   ...  Software-2018
3           17 TeraWatts  ...  Software-2018
4                   3DIQ  ...  Software-2018

[5 rows x 11 columns]

5. Replace values which have string '—' and 'Na' to np.nan for getting the missing values

In [ ]:
df_final.replace(to_replace= ['—','Na'], value= np.nan,  inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4389: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


In [ ]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2996 entries, 0 to 2995
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   company_name      2996 non-null   object 
 1   description       2996 non-null   object 
 2   website           2944 non-null   object 
 3   funding           903 non-null    float64
 4   industries        2987 non-null   object 
 5   num_employees     2987 non-null   object 
 6   headquarters      2996 non-null   object 
 7   founded_date      2996 non-null   object 
 8   industry_groups   2987 non-null   object 
 9   full_description  1968 non-null   object 
 10  filter            2996 non-null   object 
dtypes: float64(1), object(10)
memory usage: 257.6+ KB


# Filling the null values for, number of null 

1.   website: 52 null values (not a necessary feature so wont be filled)
2.   funding: 2093 null values (70% data is missing so imputation will lead to noise
3.   industries: 9 null values (manually fill) 
4.   num_employees: 9 null values (manually fill)
5.   industry_groups: 9 null values (manually fill)
6.   full_description: (to be filled if needed by task 7)




In [ ]:
df_final.head(40)

In [ ]:
df_final.to_csv("data.csv")

6. Here we download the dataset to manually fill 3,4,5

In [ ]:
df_final.iloc[df_final[df_final['industries'].isnull()].index.tolist()]['company_name']

2866                            TimeTackle
2870                            Theo Build
2899                              Parallel
2908                       Neoportals 2021
2914                             Mountains
2962                        Delivery Co-op
2969    Cognizant Microsoft Business Group
2985                                BNK-UP
2995                                  1520
Name: company_name, dtype: object

In [ ]:
df_t= pd.read_csv("data_t.csv")
df_t.drop(labels='Unnamed: 0', axis=1,inplace= True)
df_t.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2993 entries, 0 to 2992
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   company_name      2993 non-null   object 
 1   description       2993 non-null   object 
 2   website           2941 non-null   object 
 3   funding           903 non-null    float64
 4   industries        2993 non-null   object 
 5   num_employees     2993 non-null   object 
 6   headquarters      2993 non-null   object 
 7   founded_date      2993 non-null   object 
 8   industry_groups   2993 non-null   object 
 9   full_description  1968 non-null   object 
 10  filter            2993 non-null   object 
dtypes: float64(1), object(10)
memory usage: 257.3+ KB


7. checking if num_employees contain any null values post manual filling.

In [ ]:
df_t.iloc[df_t[df_t['num_employees'].isnull()].index.tolist()]['company_name']

Series([], Name: company_name, dtype: object)

8. Processing founding date string to date format and removing time

In [ ]:
print(df_t['founded_date'].unique())
import datetime
for i,date in enumerate(df_t['founded_date']):
    print(df_t.iloc[i]['founded_date'])
    df_t.loc[i, 'founded_date'] = date.replace(' 0:00', '')
    print(df_t.iloc[i]['founded_date'])
    print()
    #df_t.iloc[i]['founded_date']=date.strftime("%Y")

In [ ]:
print(df_t['founded_date'].unique())

['2018' '01-01-18' '2017' '01-01-17' '2019' '01-01-19' '2020' '01-01-20'
 '2021' '01-01-21']


In [ ]:
#make a new column to store in date time format
df_t['founded']=pd.to_datetime(df_t['founded_date'], infer_datetime_format=True, format='%Y%m%d')
df_t.drop(labels='founded_date', axis=1,inplace= True)
df_t.head()

company_name  ...    founded
0  1 Accord Technologies  ... 2018-01-01
1          1005 Ventures  ... 2018-01-01
2    1533 Systems, Inc.   ... 2018-01-01
3           17 TeraWatts  ... 2018-01-01
4                   3DIQ  ... 2018-01-01

[5 rows x 11 columns]

In [ ]:
print(df_t['founded'].unique())

['2018-01-01T00:00:00.000000000' '2017-01-01T00:00:00.000000000'
 '2019-01-01T00:00:00.000000000' '2020-01-01T00:00:00.000000000'
 '2021-01-01T00:00:00.000000000']


Since the funding column has 70% of the features missing, filling the na values will result in a lot of noise in the dataset. The data is being partitioned into 2 datasets for cleaning:


1.   Dataset with 903 rows having the funding column
2.   Dataset with 2993 rows with the funding column dropped.




In [ ]:
df_funding= df_t.dropna(subset= ['funding'])
df_funding.reset_index(inplace= True)
df_funding= df_funding.drop(labels= ['index'], axis=1)
df_funding.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903 entries, 0 to 902
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   company_name      903 non-null    object 
 1   description       903 non-null    object 
 2   website           886 non-null    object 
 3   funding           903 non-null    float64
 4   industries        903 non-null    object 
 5   num_employees     903 non-null    object 
 6   headquarters      903 non-null    object 
 7   founded_date      903 non-null    object 
 8   industry_groups   903 non-null    object 
 9   full_description  807 non-null    object 
 10  filter            903 non-null    object 
dtypes: float64(1), object(10)
memory usage: 77.7+ KB


In [ ]:
df_funding.head()

company_name  ...         filter
0         ABYA  ...  Software-2018
1       Agamon  ...  Software-2018
2        Agape  ...  Software-2018
3        Agave  ...  Software-2018
4   Agita Labs  ...  Software-2018

[5 rows x 11 columns]

In [ ]:
df_no_funding= df_t.drop(labels= ['funding'], axis=1)

In [ ]:
df_no_funding.info(), df_funding.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2993 entries, 0 to 2992
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   company_name      2993 non-null   object
 1   description       2993 non-null   object
 2   website           2941 non-null   object
 3   industries        2993 non-null   object
 4   num_employees     2993 non-null   object
 5   headquarters      2993 non-null   object
 6   founded_date      2993 non-null   object
 7   industry_groups   2993 non-null   object
 8   full_description  1968 non-null   object
 9   filter            2993 non-null   object
dtypes: object(10)
memory usage: 234.0+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903 entries, 0 to 902
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   company_name      903 non-null    object 
 1   description       903 non-null    object 
 2   

(None, None)

In [ ]:
df_no_funding.to_csv('crunchbase_with_nofunding.csv')
df_funding.to_csv('filtered_crunchbase_with_funding.csv')